smolyay/samples.py

```diff
@@ -508,3 +508,620 @@ def _create(self):
                 2 * numpy.pi * indexes / num_points
             )
         self._points = self._scale_to_domain(points, [0, 2 * numpy.pi])
+
+
+import abc
+import itertools
+
+import numpy
+import scipy.stats.qmc
+
+
+class MultidimensionalPointSet(abc.ABC):
+    """Multidimensional set of points
+
+    Creates a set of grid points within the specified domain
+
+    Parameters
+    ----------
+    domain: list
+         the domain of the grid
+    """
+
+    def __init__(self):
+        self._domain = None
+        self._points = None
+        self._valid_cache = False
+
+    @property
+    def num_dimensions(self):
+        """int: number of independent variables."""
+        return numpy.shape(self.domain)[0]
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return self._domain
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points to be sampled."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._points
+
+    def __len__(self):
+        return self.points.shape[0]
+
+    @abc.abstractmethod
+    def _create(self):
+        """Abstract method for generating the mulitdimensional grid set"""
+        pass
+
+
+class RandomPointSet(MultidimensionalPointSet):
+    """Point Set that uses randomly generated points
+
+    A Multidimensional point set that creates a grid using a Monte Carlo
+    or Quasi Monte Carlo method for generating points.
+
+    Parameters
+    ----------
+    domain: list
+        domain of the random points
+
+    number_points: int
+        number of points to generate
+
+    method: {"uniform", "halton", "sobol", "latin"}
+        the method of generating random points
+
+    seed: {int, numpy.random.Generator}
+        seed for generating the random points
+
+    options: {None, dict}, optional
+        if method != None, additional parameters passed to the QMCEngine
+    """
+
+    def __init__(self, domain, number_points, seed):
+        super().__init__()
+        self._domain = None
+        self._number_points = None
+        self._seed = None
+
+        self.domain = domain
+        self.number_points = number_points
+        self.seed = seed
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return self._domain
+
+    @domain.setter
+    def domain(self, value):
+        domain = numpy.sort(numpy.array(value, ndmin=2), axis=1)
+        if domain.ndim != 2 or domain.shape[1] != 2:
+            raise TypeError("Domain must have size (num_dimensions, 2)")
+        if any(domain[:, 0] >= domain[:, 1]):
+            raise ValueError("Lower bound must be less than upper bound")
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            self._valid_cache = False
+
+    @property
+    def number_points(self):
```

Comment:
This would be `num_points` with our naming convention.
```suggestion
    def num_points(self):
```